File: leetcode/c++/easy/_1_two_sum.cpp
// Leetcode Problem 1

#include <gtest/gtest.h>
#include <vector>
#include <unordered_map>

using namespace std;


class Solution {
public:
     vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> map;

        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i];
            if (map.find(complement) != map.end()) {
                return {map[complement], i};
            }
            map[nums[i]] = i;
        }

        return {};
    }
};

class TestTwoSum : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestTwoSum, TestCase1) {
    vector<int> nums = {2, 7, 11, 15};
    vector<int> sol = {0, 1};

    EXPECT_EQ(solution.twoSum(nums, 9), sol);
}

TEST_F(TestTwoSum, TestCase2) {
    vector<int> nums = {3, 2, 4};
    vector<int> sol = {1, 2};

    EXPECT_EQ(solution.twoSum(nums, 6), sol);
}

TEST_F(TestTwoSum, TestCase3) {
    vector<int> nums = {3, 3};
    vector<int> sol = {0, 1};

    EXPECT_EQ(solution.twoSum(nums, 6), sol);
}

File: leetcode/c#/easy/_1_TwoSum.cs
// Leetcode Problem 1

using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace easy;

public class Solution_TwoSum {
    public int[] TwoSum(int[] nums, int target) {
        Dictionary<int, int> map = [];

        for (int i = 0; i < nums.Length; i++) {
            int complement = target - nums[i];

            if (map.ContainsKey(complement)) {
                return [map[complement], i];
            }
            map[nums[i]] = i;
        }

        return [];
    }

    [TestClass]
    public class TestTwoSum {
        private Solution_TwoSum solution;

        [TestInitialize]
        public void Setup() {
            solution = new Solution_TwoSum();
        }

        [TestMethod]
        public void TestTwoSum_SimpleTest() {
            int[] nums = [2, 7, 11, 15];
            int[] expected = [0, 1];

            CollectionAssert.AreEqual(expected, solution.TwoSum(nums, 9));
        }

        [TestMethod]
        public void TestTwoSum_SimpleTest2() {
            int[] nums = [3, 2, 4];
            int[] expected = [1, 2];

            CollectionAssert.AreEqual(expected, solution.TwoSum(nums, 6));
        }

        [TestMethod]
        public void TestTwoSum_SimplgeTest3() {
            int[] nums = [3, 3];
            int[] expected = [0, 1];

            CollectionAssert.AreEqual(expected, solution.TwoSum(nums, 6));
        }
    }
}
